OIDC SSO Release - 7 September 2023 (#6450)
mendix/docs: content/en/docs/appstore/modules/oidc.md

To confirm that the authorization is working, get an access token from your Azure AD IdP and pass it to the API Endpoint using the authorization header.

#### 8.2.3 Parsing OIDC Provider Access Tokens

The OIDC SSO module provides a default access token parsing microflow to use when you are authenticating using the OIDC Provider module as your IdP.

To parse the OIDC Provider access tokens you need to do the following when performing OIDC Client Configuration: 

1. Select `OIDC.Default_OIDCProvider_TokenProcessing_CustomATP` as the **custom AccessToken processing microflow**.

    {{< figure src="/attachments/appstore/modules/oidc/enable-oidc-provider-parsing.png" >}}

2. Add the scopes `openid` and the ModelGUID or Name to the **Selected Scopes** in the OIDC Client Configuration. The ModelGUID will look something like `53f5d6fa-6da9-4a71-b011-454ec052cce8`.

> 3. Configure the user roles in your app to match the roles returned by OIDC Provider. End-users will be given the matching role when they sign into the app. If the role in the OIDC Provider token is not found in the Mendix app the end-user will not be given any user role.

Author: For SAM, we give the user a default role, `User`. Shouldn't all the parsers do the same thing to avoid confusion?

Author: Update from Sai
```suggestion
3. Configure the user roles in your app to match the roles returned by OIDC Provider. End-users will be given the matching role when they sign into the app. If the role in the OIDC Provider token is not found in the Mendix app the end-user will be given the user role `User`, but will not be given access to application.
```

>     If any one of the selected scopes of OIDC SSO matches with OIDC Provider Scopes then the user role is created. If you specify extra scopes those scopes are ignored.

Author: We have just configured the user roles, but here we are talking about **Scopes**. Should this go after the previous step, which is also about scopes?

Author: From Sai
Yes, should be moved.

Author: Moved in commit 286f4d059993c38fee9afe291d9efc387b891d97

To confirm that the authorization is working, get an access token from your Azure AD IdP and pass it to the API Endpoint using the authorization header.

#### 8.2.3 Parsing OIDC Provider Access Tokens

The OIDC SSO module provides a default access token parsing microflow to use when you are authenticating using the OIDC Provider module as your IdP.

To parse the OIDC Provider access tokens you need to do the following when performing OIDC Client Configuration: 

1. Select `OIDC.Default_OIDCProvider_TokenProcessing_CustomATP` as the **custom AccessToken processing microflow**.

    {{< figure src="/attachments/appstore/modules/oidc/enable-oidc-provider-parsing.png" >}}

2. Add the scopes `openid` and the ModelGUID or Name to the **Selected Scopes** in the OIDC Client Configuration. The ModelGUID will look something like `53f5d6fa-6da9-4a71-b011-454ec052cce8`.

    If any one of the selected scopes of OIDC SSO matches with OIDC Provider Scopes then the user role is created. If you specify extra scopes those scopes are ignored.
    

> 3. Configure the user roles in your app to match the roles returned by OIDC Provider. End-users will be given the matching role when they sign into the app. If the role in the OIDC Provider token is not found in the Mendix app the end-user will be given the user role `User`, but will not be given access to application.

Author: ```suggestion
3. Make sure that the app acting as OIDC Provider returns the right user roles for the end-users of your app. End-users will be given the matching role when they sign into the app. If the role in the OIDC Provider token is not found in the Mendix app the end-user will be given the user role `User`, but will not be given access to application.
```

Does this resolve you comment, below? 👇

Reviewer: Yes it does, thanks!

>    

Reviewer: Actually it's the other way around (referring to the first sentence):
- 1st you design your app with its userroles
- 2nd you need to ensure that the OIDC Provider assigns/provides the right roles for your app's endusers.

Suggestion:
"Make sure that the app acting as OIDC Provider returns the right roles for the end-users of your app."

>    

Reviewer: I don't understand why we are assigning the 'User' userrole but not granting access. I think we haven't agreed on such logic.
I think the behavior should be:
"If none of the roles provided in the ID-token from the OIDC Provider is found in your Mendix app, no user role will be assigned and your user will not be given access to your application".
We could say that such a user gets access as an anonymous userrole only.

Author: Will wait to hear from the team about what the behaviour here is.

Reviewer: I discussed with team today.
conclusion:
- This correctly describes current behavior
- Current behavior is consistent (across SAM and OIDC Provider AT-parsing flows)
- Team will create a ticket to assess the funvtionality and maybe define an improvement.

Long stopry short: The PR is good, this is comment is not a show stopper for release tomorrow,

The following sections describe the steps needed to make use of the ACR mechanism.

#### 8.5.1 Configuring Authentication Methods That Can Be Requested at Your IDP

To configure the ACR value(s) in the OIDC SSO module, follow these steps: 

1. Navigate to the screen where the OIDC configuration is managed. 
2. Select your client configuration and click **Edit**.
3. Add the ACR values that are supported by your IDP to the OIDC Client Configuration.

    For example, supported ACR Values for Okta IdP are: `urn:okta:loa:1fa:any` and `urn:okta:loa:2fa:any`.
 
4. Save the configuration changes. 

#### 8.5.2 Selecting the ACR Value During Sign In

When you have configured multiple ACR values for your IDP, the OIDC module shows the ACR values as additional ways to sign in on the default login page.

{{< figure src="/attachments/appstore/modules/oidc/login-acr-options.png" >}}

#### 8.5.3 Customizing the Login Page

If you want to customize this login page for your end-users, perform the following steps:

1. Create a new [page](/refguide/page/).
1. Open the App Navigation and set the newly created login page as the [Default home page](/refguide/setting-up-the-navigation-structure/#home).

> 1. Create [Role-based home pages](/refguide/setting-up-the-navigation-structure/#role-based-home-page) for the user roles. Set the target to the Home page in the module.

Author: "Set the target to the Home page in the module"

Don't we want to set the target to the new page you have just created, not an existing home page?

Author: From Sai
```suggestion
1. Create [Role-based home pages](/refguide/setting-up-the-navigation-structure/#role-based-home-page) for the user roles. Set the newly created login page as the target home page.
```

1. In the **Authentication** section, set the new login page as the **Sign-in page**.

> Depending on how your login-page works and/or which login-option is selected by the end-user, the OIDC SSO module will select the corresponding ACR value in the `acr_values` request parameter.

Author: This seems a bit vague. It doesn't really say how you should setup the new login page to ensure that the correct value is passed.

Author: From Sai
> the content of login button shows the acr value

Author: @jaapf
Is this sufficient for a customer, or is there something else that we can add. It is still a bit obscure to me, but perhaps it is clear when the customer is actually setting it up?

Reviewer: I'm not the best person to assess if this is sufficient.
In the bigger picture ACR will not be a heavily used feature, so we can allow ourselves some pragmatism.
Let's assume it is sufficinet, if we get questions/feedback, we'll do an iteration of documentation.
